### ISSUE TITLE
Searching via spatial pyramid matching kernel

### ISSUE THREAD

**Issue by asfix** [facebookresearch/faiss#359]
Hi there,

I would like to ask and learn the capability of FAISS on computing similarity based on spatial pyramid matching of histograms of feature vectors.

As is known, SPM provides a spatial representation of histogram based vectors. SPM generates a flat concataneted dense histogram based feature vectors. However,  it also encounters the diminishing coeffiecent factor to weight some of vector elements according to the level in which they were computed. In other words, the importance of some parts in large concatanated vectors is not uniformly distributed. At this point we use histogram intersection kernel to compute the similarity score between two feature vectors. 

The problem would not exist, if we had used only finest level of features since their weights are equal to 1. In this case, we can use faiss in order to compute pure L2 distance. But, what if I also want to include coarse levels? 

To sum up, is it possible to define those kinds of special similarity formulations in FAISS? What can be done to handle my specific problem? I have an idea. I may initially multiply the diminishing factor during feature vector computation. But I am not sure whether it ll be theoritically correct. It seems correct but I am not 100% sure.

So, let me reraise the main question. Is FAISS capable of defining custom similarity score computation schemes like I described above?

**Comment by mdouze** [comment]
Hi,
Faiss supports only L2 and inner product as distance functions, ie. not L1 as would be required to compute the intersection distance.
 
It may be possible to map the SPM distance computation to L2 by: 
- pre-multiplying the bins of the histogram by a factor that depends on scale
- square-rooting the descriptor components to transform L2 distance into L1 on the original vectors.

**Comment by asfix** [comment]
Thank you so much for your answer.
